paz/models/detection/frustum_pointnet.py

```diff
@@ -0,0 +1,370 @@
+
+    x = Conv2D(1024, 1, (1, 1), 'valid', activation='relu',
+               name='conv1_5')(x)
+
+    global_feat = MaxPooling2D(pool_size=[num_points, 1], padding='VALID')(x)
+
+    global_feat = tf.concat([global_feat, tf.expand_dims(tf.expand_dims(
+        one_hot_vector, 1), 1)], axis=3)
+
+    global_feat_expand = tf.tile(global_feat, [1, num_points, 1, 1])
+
+    concat_feat = tf.concat(axis=3, values=[point_feat, global_feat_expand])
+
+    x = Conv2D(512, 1, (1, 1), 'valid', activation='relu',
+               name='conv1_6')(concat_feat)
+
+    x = Conv2D(256, 1, (1, 1), 'valid', activation='relu', name='conv1_7')(x)
+
+    x = Conv2D(128, 1, (1, 1), 'valid', activation='relu', name='conv1_8')(x)
+
+    x = Conv2D(128, 1, (1, 1), 'valid', activation='relu', name='conv1_9')(x)
+
+    x = Dropout(rate=0.5)(x)
+
+    logits = Conv2D(2, 1, (1, 1), 'valid', activation=None, name='conv1_10')(x)
+
+    logits = tf.squeeze(logits, [2])  # BxNxC
+
+    return logits
+
+
+def BoxEstimationNetwork(object_point_cloud, one_hot_vec, NUM_HEADING_BIN=12,
+                         NUM_SIZE_CLUSTER=8):
+    """Amodal Box Regression network which provides Box parameters as output
+        # Arguments
+            object_point_cloud: Tensor.  Point Cloud after performing instance
+                segmentation.
+            one_hot_vector: Tensor. One hot vector of the class to which the
+                pointcloud belongs to.
+
+        # Reference
+            - [Frustum PointNet](https://arxiv.org/abs/1711.08488)
+    """
+    num_point = object_point_cloud.get_shape()[1]
+    input = tf.expand_dims(object_point_cloud, 2)
+
+    x = Conv2D(128, 1, (1, 1), 'valid', activation='relu',
+               name='conv2_1')(input)
+
+    x = Conv2D(128, 1, (1, 1), 'valid', activation='relu', name='conv2_2')(x)
+
+    x = Conv2D(256, 1, (1, 1), 'valid', activation='relu', name='conv2_3')(x)
+
+    x = Conv2D(512, 1, (1, 1), 'valid', activation='relu', name='conv2_4')(x)
+
+    x = MaxPooling2D(pool_size=[num_point, 1], padding='VALID')(x)
+
+    x = tf.squeeze(x, axis=[1, 2])
+    x = tf.concat([x, one_hot_vec], axis=1)
+
+    x = Dense(units=512, activation='relu')(x)
+    x = Dense(units=256, activation='relu')(x)
+
+    output = Dense(units=3 + NUM_HEADING_BIN * 2 + NUM_SIZE_CLUSTER * 4,
+                   activation=None)(x)
+
+    return output
+
+
+def SpatialTransformerNetwork(object_point_cloud, one_hot_vec):
+    """A Transformer Network to project object pointcloud into space
+    invariant frame
+        # Arguments
+            object_point_cloud: Tensor.  Point Cloud after performing instance
+                segmentation.
+            one_hot_vector: Tensor. One hot vector of the class to which the
+                pointcloud belongs to.
+
+        # Reference
+            - [Frustum PointNet](https://arxiv.org/abs/1711.08488)
+    """
+    num_point = object_point_cloud.get_shape()[1]
+    input = tf.expand_dims(object_point_cloud, 2)
```

Comment:
this preprocessing of shape should happen outside of the model i.e. in the preprocessing step